File: examples/go/internal/repository/memory_user_repository.go
package repository

import "examples/goapp/internal/domain"

type MemoryUserRepository struct {
	users []domain.User
}

func NewMemoryUserRepository() *MemoryUserRepository {
	return &MemoryUserRepository{
		users: []domain.User{
			domain.NewUser("1", "Ada Lovelace", "ada@example.com"),
			domain.NewUser("2", "Grace Hopper", "grace@example.com"),
		},
	}
}

func (r *MemoryUserRepository) List() []domain.User {
	return r.users
}

func (r *MemoryUserRepository) Save(user domain.User) domain.User {
	r.users = append(r.users, user)
	return user
}

File: examples/php/src/Repository/MemoryUserRepository.php
<?php

namespace App\Repository;

use App\Domain\User;

class MemoryUserRepository implements UserRepository
{
    /** @var User[] */
    private array $users = [];

    public function list(): array
    {
        return array_values($this->users);
    }

    public function save(User $user): User
    {
        $this->users[$user->getId()] = $user;
        return $user;
    }

    public function findById(string $id): ?User
    {
        return $this->users[$id] ?? null;
    }
}
